### PR TITLE
VPAID integration

### LINKED ISSUE #44: Changing "Skip" text
Hello, 

Is there any easy way to change "skip" text. 

Thanks.

### PR DESCRIPTION
**Pull request by mente**
**TL;DR - only Javascript VPAID SDKs are supported as for now. Known VPAID SDKs in Javascript - Liverail (via parameter `LR_FORMAT`, check [docs](http://support.liverail.com/publisher-integration/iab/vpaid-integration)**

Hello,

This PR adds basic support for linear VPAID ads. PR is based on squashed branch. I've kept [original branch](https://github.com/mente/videojs-vast-plugin/tree/vpaid) with all commits in if you like.

Generally what has been done:
- use forked version of vast-client to correctly detect VPAID (relies on dailymotion/vast-client-js#52)
- if VPAID is found in VAST response - use it
- small ad time indication (for VPAID only, can be applied to VAST playback as well)
- localization support for text used in plugin (closes #44). Requires video.js version 4.7.3
- tests (though I don't like it's quality)

Tested on Chrome stable latest, IE9, Firefox stable latest, iOS7.1, iOS8.0, Android 4.3 (Chrome).
PR is pretty big, looking forward to comments and ideas for improvement.

Regards,
Alex

### CODE REVIEW

**Comment by juniorplenty** [comment]
@mente trying to get this guy to work but the sample ad never runs: http://plnkr.co/edit/kvEKZ1rxIAbh7KDLWJXP?p=preview

/cc @jonhatalla

**Comment by mente** [comment]
I receive several questions like "Why this ad tag doesn't work?" 100% of them are based on Flash VPAID  SDK. I contact Actionscript developer whether it's possible to painlessly bridge javascript <-> swf SDKs. If yes - will update my findings.

**Comment by leomicheloni** [comment]
I did a basic bridge between SWF and Javascript (using external interface) if you are interested I can share the code.

**Comment by mente** [comment]
@leomicheloni would be great!

**Comment by ryanschott** [comment]
I would also be very interested in seeing that code! We are working on
something similar!

On Fri, Jan 9, 2015 at 8:40 AM, Alex Vasilenko <notifications@github.com>
wrote:

> @leomicheloni <https://github.com/leomicheloni> would be great!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/theonion/videojs-vast-plugin/pull/50#issuecomment-69335073>
> .
>



-- 

*Co-Founder, President*

ryan.schott@screenwavemedia.com <brandon.grikas@screenwavemedia.com>

*Cell Phone:* *267.371.3176*

*Skype:* schottr

*http://screenwavemedia.com/ <http://screenwavemedia.com/>*

**Comment by leomicheloni** [comment]
No problem, i'm uploading it to github in a few hours

**Comment by leomicheloni** [comment]
I couldn't find the code. But
You need to create an action script 3 file (you can use Flash Builder). Then create a class that extends Sprite

```actionscript
public class FlashAdapter extends Sprite{
pubic FlashAdapter(){
};
public LocalMethod(params: Object): void{
//this method is called from javascript
}
 
}
```

You can call javascript function from Action Script using

```actionscript
ExternalInterface.call("window.myGlobalFunction", jsonDataToSendToJavascript);
```

And you can listen to calls from javascript

ExternalInterface.addCallback( "flashMethid", this,LocalMethod );

So, you can simple load the swf object using [swfobject lib](
https://code.google.com/p/swfobject/)

Then call the method from javascript

```javascript
var swf = swfobject.flashMethod(params);
```
That's all. If you are interested we can build an adapter from scratch.

**Comment by jadbox** [comment]
Is this ready to be merged for the VPAID js part of the compatibility?

**Comment by kalail** [comment]
Would love an update on this. [Disqus](https://disqus.com/) is considering using **video.js** and **VPAID** support would help us in making that decision.

**Comment by Manbearpixel** [comment]
@kalail I too am still waiting for any more feedback about VPAID integration into this plugin. VideoJS is a great player but I am now looking for alternative players that, more than likely, will cost a licensing fee to use. I tried my luck with the VideoJS Google IMA plugin which has VPAID support, unfortunately it does not run the VPAID via Flash and failed.

**Comment by mente** [comment]
Sorry for being long on it, too much events happening in my life. Right now I don't free time for updating it and adding SWF support. If you really need it - contact me in private

**Comment by 592da** [comment]
@mente Is it money issue ? I have some website and VPAID support is critical for my site.
I will be gload to assist in order to get VPAID support for that player. I prefer this player then JW or other players.

**Comment by Manbearpixel** [comment]
@mente @592da - I too would like to contribute, if needed, to get this working. As a side note, I'm currently looking into modifying the source of the VideoJS.swf too see if I can make progress into getting VPAID working through Flash or a bridge between the two.

**Comment by pcting** [comment]
@Manbearpixel: There has been work done by the guardian around flash vpaid that i've tested and got working: https://github.com/guardian/video-js-vpaid

I'm also in the middle of refactoring code around javascript vpaid vast support that I want to contribute back to the community at some point once it's stable.

@mente: i'm refactoring your pull request and refactoring it in such a way that javascript vpaid is just another videojs tech. hopefully it works out reasonably well.

**Comment by Manbearpixel** [comment]
@pcting Oh, thanks! I had searched all over I thought for any mention of VideoJS + VPAID and did not find that fork... This is the exact thing I was working on, I'll head over and check out the project.

Cheers - mbp

**Comment by Manbearpixel** [comment]
@pcting I was looking into the video-js-vpaid repo, went through the steps to get the sandbox working and it's not running. Would you be able to provide assistance? My email is: manbearpixel@gmail.com

Thanks in advance.

**Comment by eccheung4** [comment]
@pcting hi, I also tried the video-js-vpaid repo and I get an error about "Vpaid" tech being undefined in video.dev.js when I was testing the /sandbox/videojs.html. Did you write you own plugin for the vpaid tech or did it work with the current video.dev.js?

Thanks!

**Comment by Manbearpixel** [comment]
@eccheung4 I just put together a forked repo of the guardians' VPAID attempt. My repo is a compilation of their's and some content @pcting threw my way. At this moment I do have some sort of working demonstration of Flash VPAID support. A LiveRail VPAID ad test is working, but a YuMe one is not for some reason (could be on their end). 

Anyways, check it out and run the installation on my repo and it should work fine for you. I haven't touched much since I worked on it a day or so ago since I'm waiting for YuMe to reply back with an issue.

https://github.com/Manbearpixel/video-js-vpaid

**Comment by eccheung4** [comment]
Cool, thanks @Manbearpixel ! I will check it out

**Comment by kalail** [comment]
This is interesting progress. I have a need for Flash VPAID support for Video.js in a timely manner and am willing to offer a monetary reward, pending a requirements review.

**Comment by pcting** [comment]
@eccheung4 I'm in the process of cleaning up the code to put out a proper release with an actual working example. I am hopeful I can have the code polished and ready within the next two weeks.

I unfortunately have rewritten much of the videojs-vast-plugin where I can't see it being merged back upstream.

File: example.html [hunk 1]
```diff
@@ -14,8 +14,9 @@
     }
   </style>
   <!--[if lt IE 9]><script src="lib/es5.js"></script><![endif]-->
-  <script src="http://vjs.zencdn.net/4.7.1/video.js"></script>
-  <script src="lib/videojs-contrib-ads/videojs.ads.js"></script>
+  <script src="bower_components/videojs/dist/video-js/video.js"></script>
+  <script src="bower_components/videojs-contrib-ads/src/videojs.ads.js"></script>
+  <script src="bower_components/vast-client-js/vast-client.js"></script>
 
   <script src="lib/vast-client.js"></script>
```
Comment (hacfi) [inline example.html]: Is this still necessary with line 19 using vast-client from bower components?

Comment (mente) [inline example.html]: Nope, just missed it

**Comment by eccheung4** [comment]
@pcting very cool! Looking forward to it.  I forked @Manbearpixel and was able to get LiveRail Flash, Spotx Flash, and Adaptv flash, but I couldn't get YUME working.

**Comment by Manbearpixel** [comment]
@eccheung4 Good stuff man, I'll check it out. Will be working with YuMe to determine why their integration isn't working as easily as others.

**Comment by rhekeen** [comment]
@Manbearpixel I managed to run yume vpaid with minor changes to AdContainer.as

https://github.com/rhekeen/video-js-vpaid-flash

**Comment by eccheung4** [comment]
awesome @rhekeen ! Unfortunately I tried out your repo and I get a vpaid creative ad error for your yume  vpaid test.  Is all you did was cast the loaded swf object as a Display Object? Is there anything else that maybe you forgot to commit thats needed?

thanks!

**Comment by joshmadewell** [comment]
Is this PR getting close to production ready?

**Comment by Danovadia** [comment]
any progress made in May? thx

**Comment by Manbearpixel** [comment]
@Danovadia

**Comment by Manbearpixel** [comment]
@Danovadia @joshmadewell - I'm working on a VPAID plugin for VideoJS currently. I do not have a repo on it yet since I'm still finishing it up but will source it when available. I had a working VPAID integration in the VideoJS Flash SWF but am having issues setting the source back once the ad is over, so instead I'm turning it into a plugin that will overlay the video player.

**Comment by Manbearpixel** [comment]
@Danovadia @joshmadewell - Finally wrapped up my code into a repo, feel free to check it out. Hope this helps! https://github.com/Manbearpixel/videojs-vpaid

**Comment by joshmadewell** [comment]
@Manbearpixel 

This is awesome man! Are you using this because you only use VPAID ads or do you have some kind of implementation on your end where you switch plugins based on the tag?

**Comment by Manbearpixel** [comment]
@joshmadewell 
Thanks Josh, yeah our site needed to support both VAST (HTML5) and VPAID (Flash) ads, the VAST plugin I found for VideoJS works pretty solid, there was just nothing like that for VPAID, so I had a working one but it was a tech, so it was embedded within the Flash tech and I ran into some issues. This route seemed the fastest and easiest. So now based on the ad response from our server I just switch the plugin to use.

**Comment by spra85** [comment]
@Manbearpixel revisiting this PR, does your videojs-vpaid plugin work alongside this one or meant to replace?  Curious if you're using both plugins to support both VAST and VPAID ads for your site

**Comment by joshmadewell** [comment]
@spra85 

I asked a similar question if you look at the comment just above MBP's last one. He said he implemented a way to know whether or not he needed the VAST or VPAID plugin before choosing which ad tech to use.

I needed to be able to know if the ads failed at run-time so i made some modifications to the VPAID code that MBP wrote to handle this which was very simple to do.

**Comment by Manbearpixel** [comment]
@spra85 
As noted by Josh, my plugin does NOT handle VAST. Only VPAID. Maybe down the road I can combine this VAST Plugin with my VPAID Plugin and do some sort of smart decision.
